hpx/runtime/parcelset/put_parcel.hpp

```diff
@@ -0,0 +1,246 @@
+                    std::unique_ptr<actions::continuation>, arg0_type
+                >::value
+            > is_continuation;
+
+            // Is the second paramter a action?
+            traits::is_action<arg1_type> is_action;
+
+            if (dest.get_management_type() == naming::id_type::unmanaged)
+            {
+                naming::gid_type gid = dest.get_gid();
+                naming::detail::strip_credits_from_gid(gid);
+                HPX_ASSERT(gid);
+
+                pp(detail::create_parcel::call(
+                    is_continuation, is_action,
+                    std::move(gid), std::move(addr),
+                    std::forward<Args>(args)...
+                ));
+            }
+            else if (dest.get_management_type() == naming::id_type::managed_move_credit)
+            {
+                naming::gid_type gid = naming::detail::move_gid(dest.get_gid());
+                HPX_ASSERT(gid);
+
+                pp(detail::create_parcel::call(
+                    is_continuation, is_action,
+                    std::move(gid), std::move(addr),
+                    std::forward<Args>(args)...
+                ));
+            }
+            else
+            {
+                future<naming::gid_type> splitted_gid =
+                    naming::detail::split_gid_if_needed(dest.get_gid());
+                if (splitted_gid.is_ready())
+                {
+                    pp(detail::create_parcel::call(
+                        is_continuation, is_action,
+                        splitted_gid.get(), std::move(addr),
+                        std::forward<Args>(args)...
+                    ));
+                }
+                else
+                {
+                    splitted_gid.then(
+                        hpx::util::bind(
+                            hpx::util::one_shot(
+                                [is_continuation, is_action, dest]
+                                (hpx::future<naming::gid_type> f,
+                                 typename util::decay<PutParcel>::type&& pp_,
+                                 naming::address&& addr_,
+                                 typename util::decay<Args>::type&&... args_)
+                                {
+                                    pp_(detail::create_parcel::call(
+                                        is_continuation, is_action,
+                                        f.get(), std::move(addr_),
+                                        std::move(args_)...
+                                    ));
+                                }
+                            ),
+                            hpx::util::placeholders::_1,
+                            std::forward<PutParcel>(pp), std::move(addr),
+                            std::forward<Args>(args)...
+                        )
+                    );
+                }
+            }
+        }
+
+        struct put_parcel_handler
+        {
+            void operator()(parcel&& p)
```

Comment:
The other one the file can't be const as the callback might have a non-const operator(). This one certainly can be.